File: src/JavaScript/Sorting/merge_sort.js
class MergeSort {
    /**
     * Uses Merge Sort to sort an array into ascending order.
     */
    sortAscending(array) {
        if (array.length <= 1)
            return array;
        this.mergeSortAscending(array, 0, array.length - 1);
        return array;
    }

    mergeSortAscending(array, left, right) {
        if (left >= right)
            return;
        const mid = Math.floor(left + (right - left) / 2);
        this.mergeSortAscending(array, left, mid);
        this.mergeSortAscending(array, mid + 1, right);
        this.mergeAscending(array, left, mid, right);
    }

    mergeAscending(array, left, mid, right) {
        let n1 = mid - left + 1;
        let n2 = right - mid;

        const L = new Array(n1);
        const R = new Array(n2);

        for (let i = 0; i < n1; i++)
            L[i] = array[left + i];

        for (let i = 0; i < n2; i++)
            R[i] = array[mid + 1 + i];

        let i = 0, j = 0;
        let k = left;

        while (i < n1 && j < n2) {
            if (L[i] <= R[j]) {
                array[k] = L[i];
                i++;
            } else {
                array[k] = R[j];
                j++;
            }
            k++;
        }

        while (i < n1) {
            array[k++] = L[i++];
        }

        while (j < n2) {
            array[k++] = R[j++];
        }
    }

    /**
     * Uses Merge Sort to sort an array into descending order.
     */
    sortDescending(array) {
        if (array.length <= 1)
            return array;
        this.mergeSortDescending(array, 0, array.length - 1);
        return array;
    }

    mergeSortDescending(array, left, right) {
        if (left >= right)
            return;
        const mid = Math.floor(left + (right - left) / 2);
        this.mergeSortDescending(array, left, mid);
        this.mergeSortDescending(array, mid + 1, right);
        this.mergeDescending(array, left, mid, right);
    }

    mergeDescending(array, left, mid, right) {
        let n1 = mid - left + 1;
        let n2 = right - mid;

        const L = new Array(n1);
        const R = new Array(n2);

        for (let i = 0; i < n1; i++)
            L[i] = array[left + i];

        for (let i = 0; i < n2; i++)
            R[i] = array[mid + 1 + i];

        let i = 0, j = 0;
        let k = left;

        while (i < n1 && j < n2) {
            if (L[i] >= R[j]) {
                array[k] = L[i];
                i++;
            } else {
                array[k] = R[j];
                j++;
            }
            k++;
        }

        while (i < n1) {
            array[k++] = L[i++];
        }

        while (j < n2) {
            array[k++] = R[j++];
        }
    }
}

module.exports = MergeSort

File: src/Python/Sorting/merge_sort.py
import math

class MergeSort():
    
    def sort_ascending(self, array):
        array = self.merge_sort(array, 0, len(array) - 1, True)
        return array

    def sort_descending(self, array):
        array = self.merge_sort(array, 0, len(array) - 1, False)
        return array

    def merge_sort(self, array, left, right, is_ascending):
        if (left >= right):
            return array
        mid = math.floor(left + (right - left) / 2)
        self.merge_sort(array, left, mid, is_ascending)
        self.merge_sort(array, mid + 1, right, is_ascending)
        self.merge(array, left, mid, right, is_ascending)
        return array

    def merge(self, array, left, mid, right, is_ascending):
        L = array[left : mid + 1]
        R = array[mid + 1 : right + 1]

        i = 0
        j = 0
        k = left

        while (i < len(L) and j < len(R)):
            if ((is_ascending and L[i] <= R[j]) or (not is_ascending and L[i] >= R[j])):
                array[k] = L[i]
                i += 1
            else:
                array[k] = R[j]
                j += 1
            k += 1

        while i < len(L):
            array[k] = L[i]
            i += 1
            k += 1

        while j < len(R):
            array[k] = R[j]
            j += 1
            k += 1

        return array